libnavigation-core/src/test/java/com/mapbox/navigation/core/MapboxNavigationRerouteInterruptionTest.kt

```diff
@@ -0,0 +1,91 @@
+package com.mapbox.navigation.core
+
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+import com.mapbox.navigation.core.reroute.NavigationRerouteController
+import com.mapbox.navigation.core.reroute.RerouteController
+import com.mapbox.navigation.core.reroute.RerouteState
+import io.mockk.every
+import io.mockk.mockk
+import io.mockk.verify
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.InternalCoroutinesApi
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.robolectric.ParameterizedRobolectricTestRunner
+import org.robolectric.annotation.Config
+
+@ExperimentalPreviewMapboxNavigationAPI
+@Config(shadows = [ShadowReachabilityFactory::class])
+@InternalCoroutinesApi
+@ExperimentalCoroutinesApi
+@RunWith(ParameterizedRobolectricTestRunner::class)
+internal class MapboxNavigationRerouteInterruptionTest(
+    private val previousState: RerouteState?,
+    private val expectedNumberOfInterruptions: Int,
+) : MapboxNavigationBaseTest() {
```

Comment:
maybe replace inheritance by a factory like [here](https://github.com/mapbox/mapbox-navigation-android/blob/main/libnavigation-core/src/test/java/com/mapbox/navigation/core/routerefresh/RouteRefreshControllerTest.kt#L573)? So that all default dependencies are created in factory, but if you need to override something for your test, just pass it as an optional parameter?

Reply:
I don't see why factory is better. If I use the factory, I'll need to write more code to access the created dependencies. The diff becomes larger, more copypaste is needed.
For example, with the current approach the base class introduces the coroutineRule, does all the necessary setup in before and after. Otherwise I would have to duplicate code in 2 tests. I actually think that MapboxNavigationTest can be split to more classes in the future. My approach seems to save us from writing the same code over and over.

Reply:
> I don't see why factory is better

There is no solution which is just better. They have different pros and cons. I think your solution is great, but what I'm trying to say is that maybe it would be better for us to do a different tradeoff: sacrifice something we have to get something we don't.

Factory simplifies reading and understanding of a single test. The main disadvantage of common setup is that you have to read it all to understand what's going on in one test. Every single test uses only common setup is partially. With factories, for every single test, you need to read only what the test uses. I find it easier to understand.

> The diff becomes larger, more copypaste is needed.

I'm not so sure as you. I would try and see if it really make you copy-paste 🙂

Reply:
anyway I won't insist on do such a big change, just suggested you to consider this option

Reply:
> Factory simplifies reading and understanding of a single test. The main disadvantage of common setup is that you have to read it all to understand what's going on in one test. Every single test uses only common setup is partially. With factories, for every single test, you need to read only what the test uses. I find it easier to understand.

I see your point, however, there's another side to that. I want to test just one method invocation (`setRerouteController` in my case). I've decided to move it to a separate class. First, I have to create `mapboxNavigation`. Then it turns out I have to set up a hundred other classes so that the basic initialization does not break. And when it's about such complex classes as `MapboxNavigation`, that's a lot of work. And I don't want to think about it. I just wanna call the method. With my approach I can just to that. With factories - not really.